halo2-base/README.md

```diff
@@ -0,0 +1,498 @@
+    ```rust ignore
+    QuantumCell::Existing(acell) => {
+        self.advice.push(acell.value);
+
+        if !self.witness_gen_only {
+            let new_cell =
+                ContextCell { context_id: self.context_id, offset: self.advice.len() - 1 };
+                self.advice_equality_constraints.push((new_cell, acell.cell.unwrap()));
+            }
+    }
+    ```
+- **Witness**:
+    Assigns an entirely new number into the advice column, such as a private input. When `assign_cell` is called the value is wrapped in as an `Assigned::Trivial()`` which marks it for exclusion from batch inversion.
+    ``` rust ignore
+    QuantumCell::Witness(val) => {
+        self.advice.push(Assigned::Trivial(val));
+    }
+    ```
+- **WitnessFraction**:
+    Assigns an entirely new witness value to the advice column. `WitnessFraction` exists for optimization purposes and accepts Assigned values wrapped in `Assigned::Rational()`` marked for batch inverion.
+    ``` rust ignore
+    QuantumCell::WitnessFraction(val) => {
+        self.advice.push(val);
+    }
+    ```
+- **Constant**:
+    A value that is a "known" constant. A "known" refers to known at circuit creation time to both the Prover and Verifier. When you assign a constant value there is exists another secret "Fixed" column in the circuit constraint table whose values are fixed at circuit creation time. When you assign a Constant value, you are adding this value to the Fixed column, adding the value as a witness to the Advice column, and then imposing an equality constraint between the two corresponding cells in the `Fixed` and `Advice` columns.
+``` rust ignore
+QuantumCell::Constant(c) => {
+    self.advice.push(Assigned::Trivial(c));
+        // If witness generation is not performed, enforce equality constraints between the existing cell and the new cell
+    if !self.witness_gen_only {
+    let new_cell =
+        ContextCell { context_id: self.context_id, offset: self.advice.len() - 1 };
+        self.constant_equality_constraints.push((c, new_cell));
+        }
+}
+```
+
+<br>
+
+## [GateThreadBuilder](https://github.com/axiom-crypto/halo2-lib/blob/release-0.3.0/halo2-base/src/gates/builder.rs#L49) & [GateCircuitBuilder](https://github.com/axiom-crypto/halo2-lib/blob/release-0.3.0/halo2-base/src/gates/builder.rs#L462)
+
+GateThreadBuilder tracks the cell assignments of a circuit as a `Context` allowing for the pre-computation of a circuits size and splitting the two column into multiple columns for parallelization.
+
+```rust ignore
+#[derive(Clone, Debug, Default)]
+pub struct GateThreadBuilder<F: ScalarField> {
+    /// Threads for each challenge phase
+    pub threads: [Vec<Context<F>>; MAX_PHASE],
+    /// Max number of threads
```

Comment:
didn't catch this before, but Max number of threads should say "total number of threads" (across all phases). this is just to provide new context_ids